File: backends/nest/src/adapter/paymentgw.ts
// HTTP client for the fake payment gateway, with a hard timeout so a hanging provider
// cannot hang the worker. Retry/backoff lives in the worker; circuit breaker is Phase 4.

import { PaymentGateway } from '../usecase/ports';

export class PaymentGatewayClient implements PaymentGateway {
  constructor(private readonly baseUrl: string) {
    this.baseUrl = baseUrl.replace(/\/$/, '');
  }

  async charge(orderId: string): Promise<string> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), 4000);
    try {
      const res = await fetch(`${this.baseUrl}/charges`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ order_id: orderId }),
        signal: controller.signal,
      });
      if (!res.ok) throw new Error(`payment gateway returned ${res.status}`);
      const body = (await res.json()) as { provider_ref: string };
      return body.provider_ref;
    } finally {
      clearTimeout(timer);
    }
  }
}

File: backends/fastapi/app/adapter/paymentgw.py
"""HTTP client for the fake payment gateway, with a hard timeout so a hanging provider
cannot hang the worker. Retry/backoff lives in the worker; a circuit breaker would wrap it (implemented in the Go backend).
"""
from __future__ import annotations

import httpx


class PaymentGatewayClient:
    def __init__(self, base_url: str):
        self._base_url = base_url.rstrip("/")

    async def charge(self, order_id: str) -> str:
        async with httpx.AsyncClient(timeout=4.0) as client:
            resp = await client.post(f"{self._base_url}/charges", json={"order_id": order_id})
            resp.raise_for_status()
            return resp.json()["provider_ref"]
